src/output/plugins/MacOSOutputPlugin.cxx

```diff
@@ -0,0 +1,523 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+/** This is the new CoreAudio output plugin which uses direct mode
+ *	(without Audio Unit) to output audio. Due to CoreAudio's
+ *	inner workings this means that in general all audio is converted
+ *	to float sample format. This output uses an audio converter object
+ *	to do any conversion required from MPDs input samples to the
+ *	format and channels the device asks for. The filter engine of MPD
+ *	is therefore only used for the purpose of resampling.
+ */
+
+
+/** TODO:
+ *	Callbacks for default device change and reconfiguration
+ */
+
+#include "config.h"
+#include "MacOSOutputPlugin.hxx"
+#include "../OutputAPI.hxx"
+#include "mixer/MixerList.hxx"
+#include "util/RuntimeError.hxx"
+#include "util/Domain.hxx"
+#include "util/Manual.hxx"
+#include "util/ConstBuffer.hxx"
+#include "pcm/PcmExport.hxx"
+#include <AudioToolbox/AudioConverter.h>
+#include "Log.hxx"
+#include "lib/coreaudio/CoreAudioDevice.hxx"
+#include "lib/coreaudio/CoreAudioHelpers.hxx"
+#include <boost/lockfree/spsc_queue.hpp>
+
+static constexpr Domain macos_output_domain("macos_output");
+// Set output frame buffer to double of the 512 default value
+static constexpr unsigned DEFAULT_FRAME_BUFFER_SIZE = 1024;
+// Ring buffer of at least 100ms
+static constexpr unsigned BUFFER_TIME_MS = 100;
+
+struct MacOSOutput final : AudioOutput {
+	
+	// Layer around CoreAudio
+	CoreAudioDevice device;
+
+	// Plugin settings as fetched from config
+	const char *device_name = nullptr;
+	std::vector<SInt32> channel_map;
+	bool hog_device;
+#ifdef ENABLE_DSD
+	bool dop_setting;
+#endif
+	bool integer_mode;
+	UInt32 frame_buffer_size;
+	
+	/** The following are used to do the final
+	 *	format conversion before sending the data
+	 *	to the audio device. This includes channel
+	 *	mapping and (if necessary) de-interleaving
+	 *	as well as the mandatory float conversion
+	 *	in case integer_mode is not active or not
+	 *	supported.
+	 */
+	AudioConverterRef ca_converter = nullptr;
+	AudioBufferList *out_buffer = nullptr;
+	// The format MPD sends
+	AudioStreamBasicDescription in_format;
+	// The format CoreAudio requests for IO
+	AudioStreamBasicDescription out_format;
+
+	bool pause;
+	
+	/** Required to support DoP. No
+	 *	other features currently used.
+	 */
+	Manual<PcmExport> pcm_export;
```

Comment:
If it's a DoP only feature, guard it behind a flag. Also hide it within a if (dop) switch. 
Otherwise the code will be running for all audio renderer loops. 
It's always a good idea to optimize the audio renderer loop and make it do as few things as possible.

Reply:
This runs also by default in the alsa output and I see no need to have a complex ´if´ (or `#ifdef`) logic around it for using either PcmExport or direct chunk copying. Also note that this runs inside the Play method but not in the audio render callback.

Reply:
ALSA has that running doesn't mean it's mandatory for your plugin to do so. 
The first priority in programming an audio render loop is to make sure it consumes as little CPU resources as possible, and avoid any hard to predict latency.